services/autoscaling/src/simcore_service_autoscaling/models.py

```diff
@@ -6,43 +6,79 @@
 from models_library.generated_models.docker_rest_api import Node
 
 
-@dataclass(frozen=True, kw_only=True)
-class AssignedTasksToInstance:
-    instance: EC2InstanceData
-    available_resources: Resources
-    assigned_tasks: list
-
-
-@dataclass(frozen=True, kw_only=True)
+@dataclass(kw_only=True, slots=True)
 class AssignedTasksToInstanceType:
     instance_type: EC2InstanceType
     assigned_tasks: list
+    available_resources: Resources
 
+    def has_resources_for_task(self, task_resources: Resources) -> bool:
+        return bool(self.available_resources >= task_resources)
 
-@dataclass(frozen=True)
-class AssociatedInstance:
-    node: Node
+    def assign_task(self, task, task_resources: Resources) -> None:
+        self.assigned_tasks.append(task)
+        self.available_resources -= task_resources
+
+
+@dataclass(frozen=True, kw_only=True, slots=True)
+class _BaseInstance:
     ec2_instance: EC2InstanceData
+    assigned_tasks: list = field(default_factory=list)
+    _available_resources: Resources = field(default_factory=Resources.create_as_empty)
+
+    def __post_init__(self) -> None:
```

Comment:
TIP: reuse. I can think of two possibilities
1. A mixin that inherits
2. A free function assigned to a field (as [reusable validators in pydantic](https://docs.pydantic.dev/1.10/usage/validators/#reuse-validators))

Reply:
not sure I get what you mean here @pcrespov, I am already re-using that in 2 classes.

Reply:
ok just understood. will refactor. thanks